G2-http3-protocol.md

```diff
@@ -0,0 +1,142 @@
+
+HTTP/3 stream IDs function largely the same as HTTP/2 stream IDs.
+
+#### Data frames
+
+The relationship between `DATA` frames and length prefixed messages are unchanged in HTTP/3.
+
+#### Errors
+
+HTTP/3 [has different error codes](https://quicwg.org/base-drafts/draft-ietf-quic-http.html#section-8.1) from HTTP/2. HTTP/3 errors are sent via QUIC frames instead of an `RST_STREAM` HTTP frame.
+
+HTTP/3 errors are used in three situations:
+
+* Abruptly terminating streams
+* Aborting reading of streams
+* Immediately closing HTTP/3 connections
+
+HTTP3 Code|HTTP2 Code|GRPC Code
+----------|----------|-----------
+H3_NO_ERROR(0x0100)|NO_ERROR(0)|INTERNAL - An explicit GRPC status of OK should have been sent but this might be used to aggressively [lameduck](https://landing.google.com/sre/sre-book/chapters/load-balancing-datacenter/#identifying-bad-tasks-flow-control-and-lame-ducks-bEs0uy) in some scenarios.
+H3_GENERAL_PROTOCOL_ERROR(0x0101)|PROTOCOL_ERROR(1)|INTERNAL
+H3_INTERNAL_ERROR(0x0102)|INTERNAL_ERROR(2)|INTERNAL
+H3_STREAM_CREATION_ERROR(0x0103)|n/a|INTERNAL
+H3_CLOSED_CRITICAL_STREAM(0x0104)|n/a|INTERNAL
+H3_FRAME_UNEXPECTED(0x0105)|FRAME_SIZE_ERROR|INTERNAL
+H3_FRAME_ERROR(0x0106)|FRAME_SIZE_ERROR|INTERNAL
+H3_EXCESSIVE_LOAD(0x0107)|ENHANCE_YOUR_CALM|RESOURCE_EXHAUSTED ...with additional error detail provided by runtime to indicate that the exhausted resource is bandwidth.
+H3_ID_ERROR(0x0108)|n/a|INTERNAL
+H3_SETTINGS_ERROR(0x0109)|SETTINGS_TIMEOUT(4)|INTERNAL
+H3_MISSING_SETTINGS(0x010a)|SETTINGS_TIMEOUT(4)|INTERNAL
+H3_REQUEST_REJECTED(0x010b)|REFUSED_STREAM|UNAVAILABLE - Indicates that no processing occurred and the request can be retried, possibly elsewhere.
+H3_REQUEST_CANCELLED(0x010c)|CANCEL(8)|Mapped to call cancellation when sent by a client.Mapped to CANCELLED when sent by a server. Note that servers should only use this mechanism when they need to cancel a call but the payload byte sequence is incomplete.
+H3_REQUEST_INCOMPLETE(0x010d)|n/a|INTERNAL
+H3_MESSAGE_ERROR(0x010e)|n/a|INTERNAL
+H3_CONNECT_ERROR(0x010f)|CONNECT_ERROR|INTERNAL
+H3_VERSION_FALLBACK(0x0110)|n/a|INTERNAL
+n/a|FLOW_CONTROL_ERROR(3)|INTERNAL
+n/a|STREAM_CLOSED|No mapping as there is no open stream to propagate to. Implementations should log.
+n/a|COMPRESSION_ERROR|INTERNAL
+n/a|INADEQUATE_SECURITY| PERMISSION_DENIED … with additional detail indicating that permission was denied as protocol is not secure enough for call.
+
+#### Connection management
+
+`GOAWAY` and `PING` frames exist in HTTP/3 and serve the same purpose as in HTTP/2.
+One notable difference is the `GOAWAY` frame in HTTP/2 reports the last
+successfully processed stream ID. In HTTP/3 the `GOAWAY` frame ID value must be greater
+that the last successfully processed stream ID.
+
+### Exceeding deadlines
+
+When an RPC has exceeded its deadline in HTTP/2, the server will complete the response
+with a gRPC status code of DEADLINE_EXCEEDED, then send a `RST_STREAM` frame to the
```

Comment:
Actually, we generally use a `RST_STREAM(CANCEL)` for server→client direction when the deadline expires as it is cheap and avoids flow control so any buffered data get thrown away. We generally expect that doesn't matter much because `RST_STREAM(CANCEL)` for client→server is racing with it. When our client implementations see a CANCELLED status they double-check whether the deadline has been exceeded, and if so replace the status and trailers with DEADLINE_EXCEEDED.

What you describe here is a fine server behavior, because, again, what the server does here shouldn't matter except for a race and it will resolve fine in a race. But it does keep the stream alive longer.

Reply:
Fixed